File: Send SMS/.NET/VB.NET/HTTP_example.vb

Imports System.Net
Imports System.IO
Imports System.Text
Imports System.Web

Module Module1

    Const SMS_TYPE_CLASSIC As String = "classic"
    Const SMS_TYPE_CLASSIC_PLUS As String = "classic_plus"
    Const SMS_TYPE_BASIC As String = "basic"
    Const SMS_TYPE_TEST_CLASSIC As String = "test_classic"
    Const SMS_TYPE_TEST_CLASSIC_PLUS As String = "test_classic_plus"
    Const SMS_TYPE_TEST_BASIC As String = "test_basic"

    Private Function skebbyGatewaySendSMS(
                         ByVal username As String, _
                         ByVal password As String, _
                         ByVal recipients() As String, _
                         ByVal text As String, _
                         Optional ByVal sms_type As String = "basic", _
                         Optional ByVal sender_number As String = "", _
                         Optional ByVal sender_string As String = "", _
                         Optional ByVal user_reference As String = "", _
                         Optional ByVal charset As String = "" _
                     ) As Dictionary(Of String, String)
        Dim parameters, method As String
        Dim URL As String
        Dim result As New Dictionary(Of String, String)
        Dim values() As String
        Dim temp

        URL = "http://gateway.skebby.it/api/send/smseasy/advanced/http.php"

        method = "send_sms_classic"

        Select Case sms_type
            Case SMS_TYPE_CLASSIC
                method = "send_sms_classic"
            Case SMS_TYPE_CLASSIC_PLUS
                method = "send_sms_classic_report"
            Case SMS_TYPE_BASIC
                method = "send_sms_basic"
            Case SMS_TYPE_TEST_CLASSIC
                method = "test_send_sms_classic"
            Case SMS_TYPE_TEST_CLASSIC_PLUS
                method = "test_send_sms_classic_report"
            Case SMS_TYPE_TEST_BASIC
                method = "test_send_sms_basic"
            Case Else
                method = "send_sms_classic"
        End Select

        parameters = "method=" + HttpUtility.UrlEncode(method) + "&" _
                    + "username=" + HttpUtility.UrlEncode(username) + "&" _
                    + "password=" + HttpUtility.UrlEncode(password) + "&" _
                    + "text=" + HttpUtility.UrlEncode(text) + "&" _
                    + "recipients[]=" + String.Join("&recipients[]=", recipients)

        If (sender_number <> "") And (sender_string <> "") Then
            result.Add("status", "failed")
            result.Add("code", "0")
            result.Add("message", "You can specify only one type of sender, numeric or alphanumeric")
            Return result
        End If

        If sender_number <> "" Then parameters = parameters + "&sender_number=" + HttpUtility.UrlEncode(sender_number)
        If sender_string <> "" Then parameters = parameters + "&sender_string=" + HttpUtility.UrlEncode(sender_string)
        If user_reference <> "" Then parameters = parameters + "&user_reference=" + HttpUtility.UrlEncode(user_reference)

        Select Case charset
            Case "UTF-8"
                parameters = parameters + "&charset=" + HttpUtility.UrlEncode("UTF-8")
            Case Else
        End Select

        ' Create POST
        Dim request As WebRequest = WebRequest.Create(URL)
        request.Method = "POST"
        Dim byteArray As Byte() = Encoding.UTF8.GetBytes(parameters)
        request.ContentType = "application/x-www-form-urlencoded"
        request.ContentLength = byteArray.Length
        Dim dataStream As Stream = request.GetRequestStream()
        dataStream.Write(byteArray, 0, byteArray.Length)
        dataStream.Close()

        Dim response As WebResponse
        Try
            'Trying to get the response.
            response = request.GetResponse()
        Catch ex As System.Net.WebException
            result.Add("status", "failed")
            result.Add("code", "0")
            result.Add("message", "Network error, unable to send the message")
            Return result
        End Try

        dataStream = response.GetResponseStream()
        Dim reader As StreamReader = New StreamReader(dataStream)
        Dim responseFromServer As String = reader.ReadToEnd()
        ' Clean up the streams.
        reader.Close()
        dataStream.Close()
        response.Close()

        ' Return result to calling function
        If responseFromServer.Length > 0 Then
            values = responseFromServer.Split(New Char() {"&"c})
            For Each value In values
                temp = value.Split(New Char() {"="c})
                result.Add(temp(0), temp(1))
            Next
            Return result
        End If
    End Function

    Function skebbyGatewayGetCredit(username As String, password As String, Optional charset As String = "") As Dictionary(Of String, String)
        Dim parameters, method As String
        Dim URL As String
        Dim result As New Dictionary(Of String, String)
        Dim values() As String
        Dim temp

        URL = "http://gateway.skebby.it/api/send/smseasy/advanced/http.php"
        method = "get_credit"

        parameters = "method=" + HttpUtility.UrlEncode(method) + "&" _
                   + "username=" + HttpUtility.UrlEncode(username) + "&" _
                   + "password=" + HttpUtility.UrlEncode(password)

        Select Case charset
            Case "UTF-8"
                parameters = parameters + "&charset=" + HttpUtility.UrlEncode("UTF-8")
            Case Else
        End Select

        ' Create POST
        Dim request As WebRequest = WebRequest.Create(URL)
        request.Method = "POST"
        Dim byteArray As Byte() = Encoding.UTF8.GetBytes(parameters)
        request.ContentType = "application/x-www-form-urlencoded"
        request.ContentLength = byteArray.Length
        Dim dataStream As Stream = request.GetRequestStream()
        dataStream.Write(byteArray, 0, byteArray.Length)
        dataStream.Close()

        Dim response As WebResponse
        Try
            'Trying to get the response.
            response = request.GetResponse()
        Catch ex As System.Net.WebException
            result.Add("status", "failed")
            result.Add("code", "0")
            result.Add("message", "Network error, unable to send the message")
            Return result
        End Try

        dataStream = response.GetResponseStream()
        Dim reader As StreamReader = New StreamReader(dataStream)
        Dim responseFromServer As String = reader.ReadToEnd()
        ' Clean up the streams.
        reader.Close()
        dataStream.Close()
        response.Close()

        ' Return result to calling function
        If responseFromServer.Length > 0 Then
            values = responseFromServer.Split(New Char() {"&"c})
            For Each value In values
                temp = value.Split(New Char() {"="c})
                result.Add(temp(0), temp(1))
            Next
            Return result
        End If

    End Function

    Sub Main()

        Dim result
        Dim credit_result

        ' Single dispatch
        Dim recipients(0) As String
        recipients(0) = "393471234567"

        ' Multiple dispatch
        ' Dim recipients(1) As String
        ' recipients(0) = "393471234567"
        ' recipients(1) = "393477654321"

        ' ------------ SMS Classic dispatch --------------

        ' SMS CLASSIC dispatch with custom alphanumeric sender
        result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you?", SMS_TYPE_CLASSIC, "", "John")

        ' SMS CLASSIC dispatch with custom numeric sender
        ' result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you?", SMS_TYPE_CLASSIC, "393471234567")


        ' ------------- SMS Basic dispatch ----------------
        ' result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you? By John", SMS_TYPE_BASIC)


        ' ------------ SMS Classic Plus dispatch -----------

        ' SMS CLASSIC PLUS dispatch (with delivery report) with custom alphanumeric sender
        ' result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you?", SMS_TYPE_CLASSIC_PLUS, "", "John")

        ' SMS CLASSIC PLUS dispatch (with delivery report) with custom numeric sender
        ' result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_CLASSIC_PLUS,"393471234567")

        ' SMS CLASSIC PLUS dispatch (with delivery report) with custom alphanumeric sender and custom reference string
        ' result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_CLASSIC_PLUS,"393471234567","","reference")

        ' ------------------------------------------------------------------
        '     WARNING! THE SMS_TYPE_TEST* SMS TYPES DOESN'T DISPATCH ANY SMS 
        '     USE THEM ONLY TO CHECK IF YOU CAN REACH THE SKEBBY SERVER 
        ' ------------------------------------------------------------------


        ' ------------- SMS Classic test dispatch ---------

        ' SMS CLASSIC test dispatch with custom alphanumeric sender
        ' result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC, "", "John")

        ' SMS CLASSIC test dispatch with custom numeric sender
        ' result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC,"393471234567")


        ' ------------- Testing invio SMS Classic Plus ---------

        ' SMS CLASSIC PLUS test dispatch (with delivery report) with custom alphanumeric sender
        ' result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC_PLUS, "", "John")

        ' SMS CLASSIC PLUS test dispatch (with delivery report) with custom numeric sender
        ' result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC_PLUS,"393471234567")


        ' ------------- SMS Basic test dispatch ---------------
        ' result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you? By John", SMS_TYPE_TEST_BASIC)


        If result("status") = "success" Then
            Console.WriteLine("Message Sent!")
            Console.WriteLine("Remaining SMSs: " + result("remaining_sms"))

            If (result.ContainsKey("id")) Then
                Console.WriteLine("ID: " + result("id"))
            End If
        End If

		' ------------------------------------------------------------------
		' Check the complete documentation at http:#www.skebby.com/business/index/send-docs/ 
		' ------------------------------------------------------------------
		' For eventual errors see http:#www.skebby.com/business/index/send-docs/#errorCodesSection
		' WARNING: in case of error DON'T retry the sending, since they are blocking errors
		' ------------------------------------------------------------------					
        If result("status") = "failed" Then
            Console.WriteLine("Sendign Failed")
            Console.WriteLine("Code:" + result("code"))
            Console.WriteLine("Description:" + result("message"))
        End If
        Console.ReadLine()

        ' ------------ REMAINING CREDIT Check -------------
        'credit_result = skebbyGatewayGetCredit("username", "password")

        'If credit_result("status") = "success" Then
        '    Console.WriteLine("Current Balance: " + credit_result("credit_left"))
        '    Console.WriteLine("Remaining SMS Classic: " + credit_result("classic_sms"))
        '    Console.WriteLine("Remaining SMS Basic: " + credit_result("basic_sms"))
        'End If
        'If credit_result("status") = "failed" Then
        '    Console.WriteLine("Sending request failed")
        'End If
        'Console.ReadLine()

    End Sub

End Module

File: Send-SMS/.NET/C#/HTTP_example.cs

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Net;
using System.IO;
using System.Web;


namespace ConsoleApplication1
{
    class Program
    {

        public const string SMS_TYPE_CLASSIC = "classic";
        public const string SMS_TYPE_CLASSIC_PLUS = "classic_plus";
        public const string SMS_TYPE_BASIC = "basic";
        public const string SMS_TYPE_TEST_CLASSIC = "test_classic";
        public const string SMS_TYPE_TEST_CLASSIC_PLUS = "test_classic_plus";
        public const string SMS_TYPE_TEST_BASIC = "test_basic";


        static void Main(string[] args)
        {
            // Single dispatch
            String[] recipients = new String[] { "393471234567" };

            // Multiple dispatch
            // String[] recipients = new String[] { "393471234567", "393477654321" };

            Hashtable result = new Hashtable();
            Hashtable credit_result = new Hashtable();
            String line;

            // ------------ SMS Classic dispatch --------------

            // SMS CLASSIC dispatch with custom alphanumeric sender
               result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you?", SMS_TYPE_CLASSIC, "", "John","","");

            // SMS CLASSIC dispatch with custom numeric sender
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_CLASSIC,"393471234567","","","");

            // ------------ SMS Basic dispatch --------------
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you? By John", SMS_TYPE_BASIC, "", "","","");

            // ------------ SMS Classic Plus dispatch --------------

            // SMS CLASSIC PLUS dispatch (with delivery report) with custom alphanumeric sender
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_CLASSIC_PLUS,"","John","","");

            // SMS CLASSIC PLUS dispatch (with delivery report) with custom numeric sender
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_CLASSIC_PLUS,"393471234567","","","");

            // SMS CLASSIC PLUS dispatch (with delivery report) with custom alphanumeric sender and custom reference string
            // result = skebbyGatewaySendSMS("username", "password", recipients, "Hi Mike, how are you?", SMS_TYPE_CLASSIC_PLUS, "393471234567", "", "riferimento", "");


            //  ------------------------------------------------------------------
			//     WARNING! THE SMS_TYPE_TEST* SMS TYPES DOESN'T DISPATCH ANY SMS 
			//     USE THEM ONLY TO CHECK IF YOU CAN REACH THE SKEBBY SERVER 
			//  ------------------------------------------------------------------

            // ------------- SMS Classic test dispatch ---------

            // SMS CLASSIC test dispatch with custom alphanumeric sender
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC,"","John","","");

            // SMS CLASSIC test dispatch with custom numeric sender
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC,"393471234567","","","");


            // ------------ SMS Classic Plus test dispatch -----------

            // SMS CLASSIC PLUS test dispatch (with delivery report) with custom alphanumeric sender
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC_PLUS,"","John","","");

            // SMS CLASSIC PLUS test dispatch (with delivery report) with custom numeric sender
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you?", SMS_TYPE_TEST_CLASSIC_PLUS,"393471234567","","","");

            // ------------- SMS Basic test dispatch ----------------
            // result = skebbyGatewaySendSMS("username","password",recipients,"Hi Mike, how are you? By John", SMS_TYPE_TEST_BASIC,"","","","");

            if ((string)result["status"] == "success")
            {
                Console.WriteLine("Message Sent!");
                Console.WriteLine("Remaining SMS: " + result["remaining_sms"]);

                if (result.ContainsKey("id"))
                {
                    Console.WriteLine("ID: " + result["id"]);
                }

            }
			
			// ------------------------------------------------------------------
			// Controlla la documentazione completa all'indirizzo http://www.skebby.it/business/index/send-docs/ 
			// ------------------------------------------------------------------
			// Per i possibili errori si veda http://www.skebby.it/business/index/send-docs/#errorCodesSection
			// ATTENZIONE: in caso di errore Non si deve riprovare l'invio, trattandosi di errori bloccanti
			// ------------------------------------------------------------------			
            if ((string)result["status"] == "failed")
            {
                Console.WriteLine("Sending failed");
                Console.WriteLine("Code: " + result["code"]);
                Console.WriteLine("Message: " + result["message"]);
            }
            line = Console.ReadLine();

            //// ------------ Controllo del CREDITO RESIDUO -------------
            //credit_result = skebbyGatewayGetCredit("username", "password", "");

            //if ((string)credit_result["status"] == "success")
            //{
            //    Console.WriteLine("Current Balance: " + credit_result["credit_left"]);
            //    Console.WriteLine("Remaining SMS Classic: " + credit_result["classic_sms"]);
            //    Console.WriteLine("Remaining SMS Basic: " + credit_result["basic_sms"]);
            //}

            //if ((string)credit_result["status"] == "failed")
            //{
            //    Console.WriteLine("Sending request failed");
            //}
            //line = Console.ReadLine();
        }

        private static Hashtable skebbyGatewaySendSMS(
            String username, String password,
            String[] recipients, String text,
            String sms_type, String sender_number,
            String sender_string,
            String user_reference,
            String charset)
        {
            String url = "http://gateway.skebby.it/api/send/smseasy/advanced/http.php";
            String result = "";
            String[] results, temp;
            String parameters = "";
            String method = "send_sms_classic";
            int i = 0;
            StreamWriter myWriter = null;
            HttpWebRequest objRequest = (HttpWebRequest)WebRequest.Create(url);
            objRequest.ServicePoint.Expect100Continue = false;
            
            Hashtable r = new Hashtable();

            switch (sms_type)
            {
                case SMS_TYPE_CLASSIC:
                    method = "send_sms_classic";
                    break;
                case SMS_TYPE_CLASSIC_PLUS:
                    method = "send_sms_classic_report";
                    break;
                case SMS_TYPE_BASIC:
                    method = "send_sms_basic";
                    break;
                case SMS_TYPE_TEST_CLASSIC:
                    method = "test_send_sms_classic";
                    break;
                case SMS_TYPE_TEST_CLASSIC_PLUS:
                    method = "test_send_sms_classic_report";
                    break;
                case SMS_TYPE_TEST_BASIC:
                    method = "test_send_sms_basic";
                    break;
                default:
                    method = "send_sms_classic";
                    break;
            }

            parameters = "method=" + HttpUtility.UrlEncode(method) + "&"
                         + "username=" + HttpUtility.UrlEncode(username) + "&password=" + HttpUtility.UrlEncode(password) + "&"
                         + "text=" + HttpUtility.UrlEncode(text) + "&"
                         + "recipients[]=" + string.Join("&recipients[]=", recipients);

            if (sender_number != "" && sender_string != "")
            {
                r.Add("status", "failed");
                r.Add("code", "0");
                r.Add("message", "You can specify only one type of sender, numeric or alphanumeric");
                return r;
            }

            parameters += sender_number != "" ? "&sender_number=" + HttpUtility.UrlEncode(sender_number) : "";
            parameters += sender_string != "" ? "&sender_string=" + HttpUtility.UrlEncode(sender_string) : "";

            parameters += user_reference != "" ? "&user_reference=" + HttpUtility.UrlEncode(user_reference) : "";

            switch (charset)
            {
                case "UTF-8":
                    parameters += "&charset=" + HttpUtility.UrlEncode("UTF-8");
                    break;
                default:
                    break;
            }

            objRequest.Method = "POST";
            objRequest.ContentLength = Encoding.UTF8.GetByteCount(parameters);
            objRequest.ContentType = "application/x-www-form-urlencoded";
            HttpWebResponse objResponse;
            try
            {
                myWriter = new StreamWriter(objRequest.GetRequestStream());
                myWriter.Write(parameters);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
                Environment.Exit(1);
            }
            finally
            {
                myWriter.Close();
            }
            try
            {
                objResponse = (HttpWebResponse)objRequest.GetResponse();
            }
            catch (WebException e)
            {
                r.Add("status", "failed");
                r.Add("code", "0");
                r.Add("message", "Network error, unable to send the message");
                return r;
            }
            using (StreamReader sr = new StreamReader(objResponse.GetResponseStream()))
            {
                result = sr.ReadToEnd();
                // Close and clean up the StreamReader
                sr.Close();
            }
            results = result.Split('&');
            for (i = 0; i < results.Length; i++)
            {
                temp = results[i].Split('=');
                r.Add(temp[0], temp[1]);
            }
            return r;
        }

        private static Hashtable skebbyGatewayGetCredit(
           String username, String password,
           String charset)
        {
            String url = "http://gateway.skebby.it/api/send/smseasy/advanced/http.php";
            String method = "get_credit";
            String result = "";
            String[] results, temp;
            String parameters = "";
            int i = 0;
            StreamWriter myWriter = null;
            HttpWebRequest objRequest = (HttpWebRequest)WebRequest.Create(url);
            Hashtable r = new Hashtable();

            parameters = "method=" + HttpUtility.UrlEncode(method) + "&"
                       + "username=" + HttpUtility.UrlEncode(username) + "&"
                       + "password=" + HttpUtility.UrlEncode(password);

            switch (charset)
            {
                case "UTF-8":
                    parameters += "&charset=" + HttpUtility.UrlEncode("UTF-8");
                    break;
                default:
                    break;
            }

            objRequest.Method = "POST";
            objRequest.ContentLength = Encoding.UTF8.GetByteCount(parameters);
            objRequest.ContentType = "application/x-www-form-urlencoded";
            HttpWebResponse objResponse;
            try
            {
                myWriter = new StreamWriter(objRequest.GetRequestStream());
                myWriter.Write(parameters);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
                Environment.Exit(1);
            }
            finally
            {
                myWriter.Close();
            }
            try
            {
                objResponse = (HttpWebResponse)objRequest.GetResponse();
            }
            catch (WebException e)
            {
                r.Add("status", "failed");
                r.Add("code", "0");
                r.Add("message", "Network error, unable to send the message");
                return r;
            }
            using (StreamReader sr = new StreamReader(objResponse.GetResponseStream()))
            {
                result = sr.ReadToEnd();
                // Close and clean up the StreamReader
                sr.Close();
            }
            results = result.Split('&');
            for (i = 0; i < results.Length; i++)
            {
                temp = results[i].Split('=');
                r.Add(temp[0], temp[1]);
            }
            return r;

        }
    }
}